docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,270 @@
+For example, additional software can be installed or configurations updated before the firmware is updated.
+This default execution order can also be overridden in the workflow definition, by updating the order in the `scheduled` state.
+
+The `"skip"` field is optional and the value is `false`, by default.
+It can be used to skip any operations during the development/testing phase, without fully deleting the entry from the profile.
+
+## Tedge agent handling device profile commands
+
+The `tedge-agent` handles `device_profile` commands using the workflow definition at `/etc/tedge/operations/device_profile.toml`.
+This workflow definition handles each module type using sub-operation workflows defined for that type.
+For example, the firmware module is installed by triggering a `firmware_update` sub-command
+which in turn uses the `firmware_update` workflow for that operation execution.
+Similarly software modules are installed with `software_update` subcommands and
+configuration updates are applied using `config_update` subcommands.
+These subcommands are triggered for each module defined in the profile definition in that order.
+
+Here is a sample device profile workflow:
+
+```toml
+operation = "device_profile"
+
+[init]
+action = "proceed"
+on_success = "scheduled"
+
+# Sort the inputs as desired
+[scheduled]
+script = "/etc/tedge/operations/device_profile.sh ${.payload.status} ${.payload}"
+on_success = "executing"
+on_error = { status = "failed", reason = "fail to sort the profile list"}
+
+[executing]
+operations = [
+    { operation = "firmware_update", target_operation="firmware_update" }
+    { operation = "software_update", target_operation="custom_software_update" }
+    { operation = "config_update" }
+]
+on_exec = "awaiting_sub_operation"
+on_success = "successful"
+on_error = { status = "failed", reason = "fail to apply device profile"}
+
+[awaiting_sub_operation]
+action = "await-operation-completion"
+on_success = "executing"
+on_error = { status = "failed", reason = "fail to update configuration"}
+
+#
+# End states
+#
+[successful]
+action = "cleanup"
+
+[failed]
+action = "cleanup"
+```
+
+* The workflow just proceeds to the `scheduled` state from the `init` state
+* The order of operation execution is finalized in the `scheduled` state.
+  If the `builtin` action is specified in this state, the default order as defined in the input is used.
+  This order can be modified using a `script` action, if desired.
+  The updated list is captured into a `updated_profile` field with the starting `current_index` value.
+  The `current_index` value specifies the next item in the list to be processed by the `executing` state.
+  By default, this `current_index` value starts with `0`.
+  This index value is incremented as the operation execution proceeds.
+  This index can also be used to skip the first `n` items in the list, especially when the same profile is reapplied.
```

Comment:
This raises many questions:

- Who increments this index value? The agent? A user provided script?
- When is this index value incremented?
- How is done the branching depending this index reached the end or not?

I guess that the intent is to free the user from writing the looping logic. However, as far I understand the proposal, the looping logic is scattered over different places making it hard both to implement and to understand. For instance the `"await-operation-completion"` action has to handle a looping logic even if a regular context. Furthermore, having a `current_operation` or `next_operation` is more convenient and meaningful than a loop index.

Reply:
Yeah I'm in favour of being as simple as possible, and possibly moving all as much of the logic outside of rust as it allows it to be easily modified if someone isn't happy with how we've handled it.

So we could even leave out all of the internal details about the `current_index` and it is just implementation details about how our specific script handler keeps track of which operation should be assigned to the `current_operation` fragment.

Reply:
Yeah, I agree that something like `next_operation` would be better than overloading the `await-operation-completion` action, as it is more explicit.

I used indexes instead of the individual operations themselves, for their simplicity/ease of manipulation and also to avoid the confusion with identical operation entries in the device profile list (e.g: @reubenmiller has highlighted a use-case in the past where the same `tedge` version had to be installed twice as part of a device profile: once before firmware update and once after the firmware update, if the firmware includes an older version). In such cases, I thought an index would be more deterministic in highlighting the current position on the list of operations executed.

Reply:
Moved the next operation computation logic completely into a built-in `iterator` action, which can be overridden using a script.